wagtail/core/rich_text/__init__.py

```diff
@@ -0,0 +1,51 @@
+from django.utils.safestring import mark_safe
+
+from wagtail.core.rich_text.feature_registry import FeatureRegistry
+from wagtail.core.rich_text.rewriters import EmbedRewriter, LinkRewriter, HTMLRewriter
+
+
+features = FeatureRegistry()
+
+
+# Rewriter function to be built up on first call to expand_db_html, using the utility classes
+# from wagtail.core.rich_text.rewriters along with the embed handlers / link handlers registered
+# with the feature registry
+
+FRONTEND_REWRITER = None
+
+
+def expand_db_html(html):
+    """
+    Expand database-representation HTML into proper HTML usable on front-end templates
+    """
+    global FRONTEND_REWRITER
+
+    if FRONTEND_REWRITER is None:
+        embed_rules = features.get_embed_types()
+        link_rules = features.get_link_types()
+        FRONTEND_REWRITER = HTMLRewriter([
+            LinkRewriter(link_rules), EmbedRewriter(embed_rules)
+        ])
+
+    return FRONTEND_REWRITER(html)
```

Comment:
The way I see it, `FeatureRegistry`'s responsibility is to collect together all the bits of configuration necessary for classes like `EditorHTMLConverter` to do their job. `FeatureRegistry` itself doesn't get involved in performing that conversion task - the rule / plugin objects are "black boxes" from `FeatureRegistry`'s perspective.

HTML rewriting for the front-end is a special case that doesn't fit nicely into that pattern (there's one global rewriter that's always active regardless of which rich text editor is in use, so we need to give `link_types` and `embed_types` [a proper documented spec within FeatureRegistry](https://github.com/gasman/wagtail/blob/9f48f4b9d78e4df1732361cc4cc74d55b31aa5a6/wagtail/core/rich_text/feature_registry.py#L29-L38) rather than saying "it's a black box") - but I think it makes sense to try and stick to the principle that FeatureRegistry doesn't have responsibility for the actual rewriting.

In other words: I'd say that FeatureRegistry is not the right place to implement the logic "an HTMLRewriter consists of a LinkRewriter plus an EmbedRewriter", even if all the individual components do exist there.